crypto/rsa/rsa_ossl.c

```diff
@@ -716,9 +718,25 @@ static int rsa_ossl_public_decrypt(int flen, const unsigned char *from,
                                     rsa->n, ctx))
             goto err;
 
-    if (!rsa->meth->bn_mod_exp(ret, f, rsa->e, rsa->n, ctx,
-                               rsa->_method_mod_n))
-        goto err;
+    exp = BN_get_word(rsa->e);
+    if (exp == RSA_3) {
+        BN_mod_mul_montgomery(ret, f, f, rsa->_method_mod_n, ctx);
+	    BN_mod_mul_montgomery(ret, f, ret, rsa->_method_mod_n, ctx);
+	    BN_mod_mul_montgomery(ret, ret, &rsa->_method_mod_n->RRR, rsa->_method_mod_n, ctx);
```

Comment:
nit: indentation of these last two lines is different from the line above

Reply:
Also the return values must be checked and errors appropriately handled.

Reply:
How many multiplications and how many squarings per verification operation are actually saved by this new implementation?

Reply:
Thanks for the feedback. I will update the indentation, error handling, and comments to reflect it saves one multiplication both in the case of `e=3` (4->3) and `e=2^16+1` (19->18).

Reply:
That's overselling it, no? It's not a full multiplication this PR is saving, just a reduction (`BN_from_montgomery). Let's call `MM(A, B)` to be Montgomery multiplication of A and B. Then squaring is `MM(A, A)` and just a reduction (`BN_from_montgomery`) is `MM(A, 1)`. But those are specialized, so we should expect costs to be `MM(A, 1) < MM(A, A) < MM(A, B)`.

I haven't looked carefully at `BN_mod_exp_mont`, but assuming it hasn't regressed, it should just be a sliding window exponentiation. With w = 1, which all these exponents should trigger, we get:
```
A' = MM(A, RR)  // BN_to_montgomery
Ret = MM(A', A')
Ret = MM(Ret, Ret) 15x
Ret = MM(Ret, A')
Ret = MM(Ret, 1)  // BN_fromt_montgomery
```

That's 16 squares, 2 multiplies, and 1 reduction. Then this PR is:
```
Ret = MM(A, A)
Ret = MM(Ret, Ret) 15x
Ret = MM(Ret, A)
Ret = MM(Ret, RR16)  // Recover from massive R^-1 deficit we've built up
```

That's 16 squares and 2 multiplies. I.e. you're just saving the cheapest operation, but only for reused RSA public keys (not the common case), while making Montgomery contexts more expensive across the board.

For that matter, this same savings can be obtained much, much more straightforwardly, with no new precomputations, and in a way that can generalize to any w=1, odd exponent:
```
A' = MM(A, RR)  // BN_to_montgomery
Ret = MM(A', A')
Ret = MM(Ret, Ret) 15x
Ret = MM(Ret, A)
```

That is, you just do the final multiplication against the input outside of Montgomery form, which will take the result out of Montgomery form at the same time. That is `MM(MM(AR, B), 1) = MM(A, B)`.

Reply:
I agree with @davidben that that's how the performance analysis should work out. However, based on the benchmarks posted in the PR description, it seems like `bn_mod_exp` might have a lot more overhead than we expect, and so avoiding calling `bn_mod_exp` at all might be what's accounting for a large amount of the saving. However, I would suggest that reducing the overhead of `bn_mod_exp` or creating a variant with less overhead might be a better course of action.

@davidben I did implement your suggestion in another project. Thanks!

Reply:
Totally agree with the suggestions, updated the patch, and we ran some tests internally with them already implemented. Thanks!

Reply:
Yeah, the benchmarks seem really inconsistent with saving just one Montgomery reduction. Even ignoring that reduction should cheaper be than squares and and multiplies, going from 19 to 18 operations should only give around a 5% speedup, not 21%. I would expect some efficiency loss from going through a much more general windowed multiplication, but nothing that much. (When I hacked it up on our end, it was only an 8% speedup, which is a bit more consistent.)

I agree with Brian that something strange is likely going on with `bn_mod_exp`. I think it would be better for OpenSSL to invest in that before accepting this PR. Specializing code can be worthwhile, but it doesn't make sense to pay that cost yet when we have evidence something is off with the generic code anyway.